### ISSUE TITLE
Move TypeSet TestCheckFunc from AWS Provider repository to SDK

### ISSUE THREAD

**Issue by anGie44** [hashicorp/terraform-plugin-sdk#595]
### SDK version

```
v2.0+
```

### Use-cases

* With the release of `v2.0` came the need to substitute current acceptance test checks that used a pre-calculated hash in the `key` e.g. 	`resource.TestCheckResourceAttr(resourceName, "parameter.1708034931.name", "character_set_results")`.
* With the work accomplished in https://github.com/terraform-providers/terraform-provider-aws/pull/13556, it would be great to make this a part of the SDK acceptance testing framework and accessible to any provider who upgrades to `v2.0`+ of the SDK. 

### Attempted Solutions

* Within the AWS Provider:
  * aws/internal/tfawsresource/testing.go currently holds `TestCheckTypeSet*` methods

### References

* https://github.com/terraform-providers/terraform-provider-aws/pull/13556

**Comment by dak1n1** [comment]
I've been working on migrating the Kubernetes provider to SDKv2. I originally used a copy of the `tfawsresource` package from PR https://github.com/terraform-providers/terraform-provider-aws/pull/13556 , but I found it actually wasn't needed. I could be wrong, but I found SDK v2 to be capable of natively supporting the same features as that custom package. I'll give an example.

Our SDK v1 test initially gave some errors after I upgraded the import paths to use SDK v2:

```
=== RUN   TestAccKubernetesEndpoints_basic
    resource_kubernetes_endpoints_test.go:22: Step 1/4 error: Check failed: 6 errors occurred:
                * Check 8/14 error: Error in test check: kubernetes_endpoints.test: Attribute 'subset.1675725831.address.#' not found
        Test check address "subset.1675725831.address.#" likely indexes into TypeSet
        This is currently not possible in the SDK
```

Initially, I googled the error and found this comment describing the problem. https://github.com/hashicorp/terraform-plugin-sdk/issues/196#issuecomment-649121248

I used the custom package listed in that comment (tfawsresource), and rewrote the checks to look like this:

```
tfk8sresource.TestCheckTypeSetElemNestedAttrs("kubernetes_endpoints.test", "subset.*.address.*", map[string]string{"ip": "10.0.0.5"}),
tfk8sresource.TestCheckTypeSetElemNestedAttrs("kubernetes_endpoints.test", "subset.*.not_ready_address.*", map[string]string{"ip": "10.0.0.10"}),
tfk8sresource.TestCheckTypeSetElemNestedAttrs("kubernetes_endpoints.test", "subset.*.not_ready_address.*", map[string]string{"ip": "10.0.0.11"}),
tfk8sresource.TestCheckTypeSetElemNestedAttrs("kubernetes_endpoints.test", "subset.*.port.*", map[string]string{
        "name":     "httpstransport",
        "port":     "443",
        "protocol": "TCP",
}),
tfk8sresource.TestCheckTypeSetElemNestedAttrs("kubernetes_endpoints.test", "subset.*.port.*", map[string]string{
        "name":     "httpstransport2",
        "port":     "444",
        "protocol": "TCP",
}),
tfk8sresource.TestCheckTypeSetElemNestedAttrs("kubernetes_endpoints.test", "subset.*.port.*", map[string]string{
        "name":     "aaaa",
        "port":     "442",
        "protocol": "TCP",
}),
```

This worked and I was happy with how readable it is. It still might be worth adding to the SDK. However, I found that it was not necessary to use this library, because this syntax also worked natively with the SDKv2:

```
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.#", "2"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.%", "3"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.address.#", "1"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.address.0.%", "3"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.address.0.ip", "10.0.0.5"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.#", "1"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.0.name", "httptransport"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.0.port", "82"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.0.protocol", "TCP"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.#", "3"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.1.name", "httpstransport"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.1.port", "443"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.1.protocol", "TCP"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.2.name", "httpstransport2"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.2.port", "444"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.2.protocol", "TCP"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.0.name", "aaaa"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.0.port", "442"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.port.0.protocol", "TCP"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.address.#", "2"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.address.0.%", "3"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.address.0.ip", "10.0.0.7"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.address.1.%", "3"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.address.1.ip", "10.0.0.6"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.address.1.hostname", "test-hostname"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.not_ready_address.0.%", "3"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.not_ready_address.0.ip", "10.0.0.10"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.not_ready_address.1.ip", "10.0.0.11"),
resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.1.not_ready_address.1.%", "3"),
```

It's a bit harder to read, but it does allow me to check every field of our TypeSet schema elements. Here's the same information represented in a Terraform config:

```
resource "kubernetes_endpoints" "test" {
  metadata {
    name = "%s"
  }

  subset {
    address {
      ip = "10.0.0.5"
    }

    port {
      name     = "httptransport"
      port     = 82
      protocol = "TCP"
    }
  }

  subset {
    address {
      ip        = "10.0.0.6"
      hostname  = "test-hostname"
      node_name = "test-nodename"
    }

    address {
      ip = "10.0.0.7"
    }

    not_ready_address {
      ip = "10.0.0.10"
    }

    not_ready_address {
      ip = "10.0.0.11"
    }
    port {
      name     = "httpstransport"
      port     = 443
      protocol = "TCP"
    }

    port {
      name     = "httpstransport2"
      port     = 444
      protocol = "TCP"
    }

    port {
      name     = "aaaa"
      port     = 442
      protocol = "TCP"
    }
  }
}
```

I found I was able to check all of those fields, even though they're TypeSet in the schema. I hope this is helpful information for others who are upgrading to SDKv2. Here's my schema, just to verify that it does contain TypeSet:

```
func resourceKubernetesEndpoints() *schema.Resource {
  return &schema.Resource{
    CreateContext: resourceKubernetesEndpointsCreate,
    ReadContext:   resourceKubernetesEndpointsRead,
    UpdateContext: resourceKubernetesEndpointsUpdate,
    DeleteContext: resourceKubernetesEndpointsDelete,
    Importer: &schema.ResourceImporter{
      StateContext: schema.ImportStatePassthroughContext,
    },

    Schema: map[string]*schema.Schema{
      "metadata": namespacedMetadataSchema("endpoints", true),
      "subset": {
        Type:        schema.TypeSet,
        Description: "Set of addresses and ports that comprise a service. More info: https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors",
        Optional:    true,
        Elem:        schemaEndpointsSubset(),
        Set:         hashEndpointsSubset(),
      },
    },
  }
}

func schemaEndpointsSubset() *schema.Resource {
  return &schema.Resource{
    Schema: map[string]*schema.Schema{
      "address": {
        Type:        schema.TypeSet,
        Description: "IP address which offers the related ports that are marked as ready. These endpoints should be considered safe for load balancers and clients to utilize.",
        Optional:    true,
        MinItems:    1,
        Elem:        schemaEndpointsSubsetAddress(),
        Set:         hashEndpointsSubsetAddress(),
      },
      "not_ready_address": {
        Type:        schema.TypeSet,
        Description: "IP address which offers the related ports but is not currently marked as ready because it have not yet finished starting, have recently failed a readiness check, or have recently failed a liveness check.",
        Optional:    true,
        MinItems:    1,
        Elem:        schemaEndpointsSubsetAddress(),
        Set:         hashEndpointsSubsetAddress(),
      },
      "port": {
        Type:        schema.TypeSet,
        Description: "Port number available on the related IP addresses.",
        Optional:    true,
        MinItems:    1,
        Elem:        schemaEndpointsSubsetPort(),
        Set:         hashEndpointsSubsetPort(),
      },
    },
  }
}
```
If needed, you can browse more of the source in my SDKv2 upgrade PR here: https://github.com/hashicorp/terraform-provider-kubernetes/pull/1027

Also I wanted to give a big credit to the author of `tfawsresource`. Without that, I wouldn't have been able to view the names of all the fields to check in my config. The debug output from it was very valuable!

### TL;DR

Instead of using the SDK v1 path names like this:

```
subset.1675725831.port.3819788156.protocol
```

Use the SDK v2 path names like this:

```
subset.0.port.0.protocol
```
